capsules/extra/src/at24c_eeprom.rs

```diff
@@ -0,0 +1,259 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2023.
+
+//! Driver for the AT24C32/64 EEPROM memory. Built on top of the I2C interface.
+//! Provides interface for the NonvolatileToPages driver.
+//!
+//! Datasheet:
+//! <https://ww1.microchip.com/downloads/en/devicedoc/doc0336.pdf>
+//!
+//! > The AT24C32/64 provides 32,768/65,536 bits of serial electrically erasable and programmable
+//! > read only memory (EEPROM) organized as 4096/8192 words of 8 bits each. The device’s cascadable
+//! > feature allows up to 8 devices to share a common 2- wire bus. The device is optimized for use
+//! > in many industrial and commercial applica- tions where low power and low voltage operation are
+//! > essential. The AT24C32/64 is available in space saving 8-pin JEDEC PDIP, 8-pin JEDEC SOIC,
+//! > 8-pin EIAJ SOIC, and 8-pin TSSOP (AT24C64) packages and is accessed via a 2-wire serial
+//! > interface.
+//!
+//! Usage
+//! -----
+//!
+//! ```rust
+//! let i2cmux = I2CMuxComponent::new(i2c0, None).finalize(components::i2c_mux_component_static!());
+//!
+//! let at24c_buffer = static_init!([u8; 34], [0; 34]);
+//!
+//! let at24c_i2c_device = static_init!(I2CDevice, I2CDevice::new(i2cmux, 0x50));
+//! let at24c_capsule = static_init!(capsules_extra::at24c_eeprom::AT24C,capsules_extra::at24c_eeprom::AT24C::new(
+//!             at24c_i2c_device,
+//!             at24c_buffer,
+//!         ) );
+//! at24c_i2c_device.set_client(at24c_capsule);
+//!
+//! let nonvolatile_storage = components::nonvolatile_storage::NonvolatileStorageComponent::new(
+//!         board_kernel,
+//!         capsules_extra::nonvolatile_storage_driver::DRIVER_NUM,
+//!         at24c_capsule,
+//!         0x0,
+//!         0x10000,
+//!         0x0,
+//!         0x0,
+//!     ).finalize(components::nonvolatile_storage_component_static!(capsules_extra::at24c_eeprom::AT24C));
+//! ```
+
+use core::cell::Cell;
+use core::cmp;
+
+use kernel::hil::i2c::{Error, I2CClient, I2CDevice};
+use kernel::utilities::cells::{OptionalCell, TakeCell};
+use kernel::{hil, ErrorCode};
+
+const PAGE_SIZE: usize = 32;
+
+pub struct EEPROMPage(pub [u8; PAGE_SIZE]);
+
+impl Default for EEPROMPage {
+    fn default() -> Self {
+        Self([0; PAGE_SIZE])
+    }
+}
+
+impl AsMut<[u8]> for EEPROMPage {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+enum State {
+    Idle,
+    Reading,
+    Writing,
+    Erasing,
+}
+
+pub struct AT24C<'a> {
+    i2c: &'a dyn I2CDevice,
+    buffer: TakeCell<'static, [u8]>,
+    client_page: TakeCell<'a, EEPROMPage>,
+    flash_client: OptionalCell<&'a dyn hil::flash::Client<AT24C<'a>>>,
+    state: Cell<State>,
+}
+
+impl<'a> AT24C<'a> {
+    pub fn new(i2c: &'a dyn I2CDevice, buffer: &'static mut [u8]) -> Self {
```

Comment:
Does `buffer` always need to be `PAGE_SIZE+2` here? If so, does `buffer: &'static mut [u8; PAGE_SIZE+2]` compile? Similarly the `TakeCell` can hold a `[u8; PAGE_SIZE+2]`.

I know that works with hard-coded numbers, as other things in Tock do it: https://github.com/tock/tock/blob/master/kernel/src/process_checker/basic.rs#L119

Reply:
Yes, buffer always need to hold `PAGE_SIZE+2` bytes, `PAGE_SIZE` for the data, 2 bytes for addressing